### ISSUE TITLE
Ansible version shipped with Uyuni Tools for openSUSE/SLES is too old for modern playbooks

### ISSUE THREAD

**Issue by stdevel** [uyuni-project/uyuni#7532]
# Short description

The Ansible version (2.9.27) shipped with Uyuni Tools is too old (September 2021).

# Details

The most current playbooks won't work with the particular version shipped with Uyuni Tools for openSUS/E/SLES (2.9.27). It doesn't support Ansible collections and also FQCN (Fully-qualified collection names) for modules aren't working.

Are there any plans to update the Ansible version? Can't find any hints in the documentation about this.

I also raised a SUSE support case for SUSE Manager 4.3 for a customer.

# Acceptance criteria

- [ ] Research options to upgrade to more recent Ansible version

**Comment by juliogonzalez** [comment]
Some things to keep in mind:
- Check that there are no breaking changes in terms of the syntax of playbooks (if there are any, we'll need to warn about it)
- Seems ALP have this a workload: https://build.opensuse.org/project/show/SUSE:ALP:Workloads and it seems the workload just takes ansible (ansible-core) from Tumbleweed: https://build.opensuse.org/package/view_file/SUSE:ALP:Workloads/ansible-container/_service:obs_scm:Dockerfile?expand=1

**Comment by meaksh** [comment]
UPDATE: After some research, we cannot easily update our Ansible version to latest `ansible-core` version.

Reason: Latest `ansible-core` version are not compatible with Python 3.6 anymore. Last version with Python 3 support was 2.11 (already EOL).

We have the new Python stack (3.11) available for SLE15SP4+ / Leap 15.4+, so maybe building `ansible-core` for Python 3.11 could be an option, but not in our case, as we provide `ansible` package via SLE15/Leap15 Client Tools, and those channels builds on SLE-15 GA / Leap 15.0.

I did some initial experiments to check whether `ansible-core` is really using functionality which is not available in Python 3.6, and there seems to be indeed some problems:

- Import errors due missing `annotations` module in `__futures__`
- Type errors due incompatible syntax.
- dataclasses missing.

I keep exploring a bit whether using `ansible-core` 2.15 with Python 3.6 is really feasible (probably not desired). I'll also provide some notes about which options / alternatives we have.

Interesting URLs:
- [ansible-core support matrix](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#ansible-core-support-matrix)

**Comment by vzhestkov** [comment]
`Alternative 2` sounds better for me, 5 also could work as well, but the others either keeps outdated ansible or push us to way more maintanance. `Alternative 1` scares me :-D

**Comment by agraul** [comment]
> These are some of the alternatives that we could take:
> ### Alternative 1: Check if `ansible-core` 2.15 or 2.16 really works with Python 3.6 and push update to SLE/Leap15 Client Tools.
> 
>     * Seems not really trivial. Core is using functionality which is not in Python 3.6:
>       
>       * Import errors due missing annotations module in **futures**
>       * Type errors due incompatible syntax.
>       * dataclasses missing.
>       * Requires Python dependencies also not compatible with Python 3.6
> 
That's a lot of backporting work to be done. An uphill battle we should avoid.

> 
> ### Alternative 2: Upgrade `ansible-core` to latest version based on Python 3.11 in our SLE/Leap15 Client Tools.
> 
>     * Requires "SLE-Module-Python3", only available for 15.4+.
> 
>     * If you are, i.a., in SLE15SP2 and you want a modern "ansible", this is not possible. Switch to next SP.
> 
>     * We could even set EOL for old ansible 2.9 based on Python 3.6 in our SLE/Leap15 Client Tools at some point.
> 
>     * Could maybe cause installation problems if Python 3.11 is not available in the subscribed channels for the client.
> 

Our Ansible integration is only meant for SLE 15 SP3+. We don't need to care about SLE 15 SP2, this is a good option. I do wonder about SLE 15 SP3 LTSS though, maybe this effort would bump the SLE requirement to SP4+. That trade-off might be worth it though.

We need to think about how to establish the dependency between the Python3 module and the ansible role, but I guess that shouldn't be too hard to implement.

> 
> ### Alternative 3: We provide modern Python 3.11 in our SLE/Leap15 Client Tools + modern Ansible and deps.
> 
>     * We would need to make Python 3.11, ansible and all necessesary dependencies to build in SLE15SP0.
> 
>     * Packages could linked to "SLE-Module-Python3" modules.
> 
>     * Providing all those packages in the client tools only for "ansible" is maybe too much.
> 
>     * More packages to maintain in our pipeline.
> 
I don't think we have bandwidth to duplicate a lot of packaging. And all that just to have a newer Ansible version available on SLE 15 SP3 client? I think Alternative 2 is strictly better.

> 
> ### Alternative 4: We do not upgrade `ansible` and keep 2.9 in our SLE/Leap15 Client Tools.
> 
>     * SLE15 is based on Python 3.6, so only "ansible" 2.9 will be shipped in Client Tools.
> 
>     * A more modern Python / ansible can be provided via some other channel, i.a. (PackageHub, ...).
> 
>     * If you are, i.a., in SLE15SP2 and you want a modern "ansible", this is not possible. Switch to next SP.
> 
Staying on an old Ansible version will harm us. The Ansible Integration won't be useful if users can't use it for off-the-shelf Ansible content.

> 
> ### Alternative 5: We provide a venv-ansible in the SLE/Leap15 Client Tools
> 
>     * We would provide an Ansible Bundle like we do with Salt: Python + Ansible + Dependencies
> 
>     * This is probably something we don't want to do at this point. Lot of work expected.
> 
This is the same as Alternative 3, just hiding the dependencies in the venv instead of having them in the repo. I don't like the amount of work this solution requires. Alternative 2 sounds better.

> 
> ### Alternative 6: We provide a modern Ansible via container instead of RPM.
> 
>     * We do not upgrade "ansible" in our SLE/Leap15 Client Tools.
> 
>     * We don't care about the Python version in the system as we provide our own Python 3.11 and deps inside the container.
> 
>     * The "SLE-Module-Containers" will be required or we include container runtime depdencies (docker/podman/runc) in the SLE/Leap15 Client Tools.
> 
This could be an option in the future. Having containerized workloads is something Uyuni will support eventually, but we're not there yet. As of now, we don't have a mechanism to sync the container image.

This is, afaik, ALPs approach when it comes to Ansible, I expect us to support this eventually.

To summarize: Alternative 2, maybe with a minimal SLE requirement from SLE 15 SP3 to SLE 15 SP4+Python Module, sounds like a good approach. We'll get to concentrate on the core (making a recent ansible version available) without having to maintain all dependencies.

**Comment by m-czernek** [comment]
My (perhaps naive) 2c is that relying on the host system for Python and dependencies is better avoided for modern development (I have run into so many issues just because we still use Python 3.6).

Consequently, to me, moving to some isolating environment, like containers or venvs, is the preferable move. I would prefer containers, but I have no inherent problem with RPMs - only Python RPMs that do not work in isolation.

As a side note, it seems to me that with the current containerization efforts that Uyuni is going through, having Uyuni's automation system be modular and independent of the monolithic base code would be an advantage. IIUIC, the containerization effort aimed at breaking Salt out of the container as its next move. 

Outside of that, I agree that alternative 2 seems like the best candidate in short term. To me, personally, it doesn't solve the underlying issue of isolation, so I view it as bringing more technical debt in the future. But, it does fit with the current deployment strategy that we use for Uyuni.

**Comment by ycedres** [comment]
Agreed with the general feeling that `Alternative 2` is pretty reasonable but also keeping an eye and somehow take steps to reach `Alternative 6` in the near future.

**Comment by agraul** [comment]
> My (perhaps naive) 2c is that relying on the host system for Python and dependencies is better avoided for current development (I have run into so many issues just because we still use Python 3.6).
> 
> Consequently, to me, moving to some isolating environment, like containers or venvs, is the preferable move. I would prefer containers, but I have no inherent problem with RPMs - only Python RPMs that do not work in isolation.
> 
I feel your pain and I agree with you. Isolated environments are easier to handle. The problem with `venv`, at least how we did it for Salt, is that we had to duplicate all the RPMs and create our own flavor for each, just to include it into the `venv`. That does not scale, especially not because we manually have to rebase these RPMs regularly.

I also prefer a container and I'm pretty positive that we will move to it once we have the infrastructure ready. It's how ALP offers ansible (https://build.opensuse.org/package/show/SUSE:ALP:Workloads/ansible-container) and we will need to support it.

> As a side note, it seems to me that with the current containerization efforts that Uyuni is going through, having Uyuni's automation system be modular and independent of the main base code would be an advantage. IIUIC, the containerization effort aimed at breaking Salt out of the container as its next move. Perhaps we can do that regardless though, with or without isolating packaging format.
> 

The ansible part is already "modular" in that the control node is a different machine than the Uyuni server. The flow goes "Uyuni -> salt ansible-control-node ansible.* -> ansible-control-node: ssh to controlled machines". We just need to be able to call `ansible-*` on the control node.


> Outside of that, I agree that alternative 2 seems like the best candidate in short term. To me, personally, it doesn't solve the underlying issue of isolation, so I view it as bringing more technical debt in the future. But, we might be solving the isolation problem up the stack (e.g. in the whole container as opposed to isolating each part of the Uyuni system) and it does fit with the current deployment strategy that we use for Uyuni.

Alternative 2 is basically the same solution we currently have, except that it requires the reduced Python stack from the Python module. I don't see it as adding more tech debt, but it surely does not reduce it.

When it comes to solution 6, it's just the fact that we can't mirror the image as of now. Deploying a container won't be a problem, calling `podman run` is already taken care of by the ALP ansible container. We can re-use that approach as well. Now, for SUMA 5.0, to be released next summer, something to mirror containers is hopefully available. I think the proxy will only be available in the containerized form.

**Comment by meaksh** [comment]
After having some discussions with @juliogonzalez, it seems the "Alternative 2" would be a bit problematic, mainly because:

- Leap/SLE15 Client Tools needs to build based on 15SP0 (or SP1), and Python3.11 is only available for 15.4+.
- We cannot make this client tools channel to depend on "SLE-Module-Python3".
- If you are in, i.a. in 15SP2, zypper will tell you have an upgrade for "ansible" but will fail to install due missing Python 3.11.
- Having a package in a channel which is not installable, as it would happen if "SLE-Module-Python3" is not available or enabled, will lead to installation problems and probably issues at Maintenance team side.
- We would need to build Python3.11 for 15SP0 (SP1) and ship it also in the Client Tools, so basically Alternative 3.

OTOH we have readapted "Alternative 4" to a more feasible approach, which will probably fits better than "Alternative 2" due the above mentioned problems.

This would be the new approach (also edited in the previous coment)

### Alternative 4: Provide a new "ansible" out of SLE/Leap15 Client Tools:
- SLE/Leap15 Client tools are based on 15SP0 (or SP1), so only Python 3.6 available.
- No update for "ansible" in SLE15/Leap15 Client Tools. It sticks to 2.9 version.
- We will provide new "ansible" version in a new channel for SLE/Leap 15.4+.
- For SLE, as the new "ansible" would require on Python 3.11, we would provide the new ansible in a new child channel from "SLE-Module-Python3" for SLE15SP4+.
- For Leap, the new "ansible" will come on a new child channel for the Leap Client Tools for 15.4+.

**Comment by meaksh** [comment]
FYI here we have "ansible" building for 15.4 based on Python 3.11: https://build.opensuse.org/project/show/home:PSuarezHernandez:branches:systemsmanagement:ansible

There are few dependencies I needed to add (not included in "SLE-Module-Python3"):
- python-commentjson (not even in SLE)
- python-resolvelib (not even in SLE)
- python-jmespath 
- python-botocore

NOTE: As this was just a POC, I just took the deps from OBS devel projects but we should be taking from SLE (prefered) or Factory.

**Comment by meaksh** [comment]
After a meeting between Ion Squad, RelEngs and PO, we have decided to go with proposed "Alternative 4".

NOTE: Once we have the new ansible version released, we should consider dropping support for old 2.9 version at some point, as it is still available in the client tools for LTSS versions (SLE15SP3 LTSS ends on 31 Dec 2025).

I will prepare an epic card to track all necessary work for this.

**Comment by meaksh** [comment]
The EPIC card has been created and reviewed by RelEngs: https://github.com/SUSE/spacewalk/issues/23281

It was now sent to Planning board. All done here for now, closing issue.
